File: slizaa-hierarchicalgraph/graphdb/org.slizaa.hierarchicalgraph.graphdb.mapping.cypher/src/main/java/org/slizaa/hierarchicalgraph/graphdb/mapping/cypher/AbstractQueryBasedHierarchyProvider.java
package org.slizaa.hierarchicalgraph.graphdb.mapping.cypher;

import org.slizaa.core.boltclient.IBoltClient;
import org.slizaa.hierarchicalgraph.graphdb.mapping.spi.IHierarchyDefinitionProvider;
import org.slizaa.hierarchicalgraph.graphdb.mapping.spi.ParentChildNode;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

public abstract class AbstractQueryBasedHierarchyProvider implements IHierarchyDefinitionProvider, IBoltClientAware {

  private List<Long> _toplevelNodeIds;

  private List<ParentChildNode> _parentChildNodes;


  @Override
  public void initialize(IBoltClient boltClient) throws Exception {

    checkNotNull(boltClient);

    this._toplevelNodeIds = new ArrayList<>();
    for (String query : toplevelNodeIdQueries()) {
      this._toplevelNodeIds.addAll(
          boltClient.asyncExecCypherQueryAndTransformResult(query, result -> result.list(r -> r.get(0).asLong())).get());
    }

    this._parentChildNodes = new ArrayList<>();
    for (String query : parentChildNodeIdsQueries()) {
      this._parentChildNodes.addAll(
          boltClient.asyncExecCypherQueryAndTransformResult(query, result -> result.list(r -> new ParentChildNode(
              r.get(0).asLong(),
              r.get(1).asLong(),
              r.get(2).asString()
          ))).get());
    }
  }

  @Override
  public List<Long> getToplevelNodeIds() throws Exception {
    return this._toplevelNodeIds;
  }

  @Override
  public List<ParentChildNode> getParentChildNodeIds() throws Exception {
    return this._parentChildNodes;
  }

    protected abstract String[] toplevelNodeIdQueries();

    protected abstract String[] parentChildNodeIdsQueries();
}

File: hierograph-hierarchicalgraph/io.hierograph.hierarchicalgraph.graphdb.mapping.service/src/main/kotlin/io/hierograph/hierarchicalgraph/graphdb/mapping/service/AbstractQueryBasedHierarchyProvider.kt
/*
 * Copyright 2026 Gerd Wuetherich
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.hierograph.hierarchicalgraph.graphdb.mapping.service

import io.hierograph.hierarchicalgraph.graphdb.mapping.spi.IHierarchyDefinitionProvider
import io.hierograph.hierarchicalgraph.graphdb.mapping.spi.ParentChildNodeId
import io.hierograph.hierarchicalgraph.graphdb.mapping.spi.ToplevelNodeId
import io.hierograph.boltclient.IBoltClient

abstract class AbstractQueryBasedHierarchyProvider : IHierarchyDefinitionProvider, IBoltClientAware {

    private var _toplevelNodeIds: List<ToplevelNodeId> = emptyList()
    private var _parentChildNodes: List<ParentChildNodeId> = emptyList()

    override fun initialize(boltClient: IBoltClient) {
        val toplevel = mutableListOf<ToplevelNodeId>()
        for (query in toplevelNodeIdQueries()) {
            val results = boltClient.asyncExecCypherQueryAndTransformResult(query) { result ->
                result.list { r ->
                    ToplevelNodeId(
                        id = r.get(0).asLong(),
                        kind = parseKind(r.get(1).asString())
                    )
                }
            }.get()
            toplevel.addAll(results)
        }
        _toplevelNodeIds = toplevel

        val parentChild = mutableListOf<ParentChildNodeId>()
        for (query in parentChildNodeIdsQueries()) {
            val results = boltClient.asyncExecCypherQueryAndTransformResult(query) { result ->
                result.list { r ->
                    ParentChildNodeId(
                        parentId = r.get(0).asLong(),
                        childId = r.get(1).asLong(),
                        childKind = parseKind(r.get(2).asString())
                    )
                }
            }.get()
            parentChild.addAll(results)
        }
        _parentChildNodes = parentChild
    }

    override fun getToplevelNodeIds(): List<ToplevelNodeId> = _toplevelNodeIds

    override fun getParentChildNodeIds(): List<ParentChildNodeId> = _parentChildNodes

    protected abstract fun toplevelNodeIdQueries(): Array<String>

    protected abstract fun parentChildNodeIdsQueries(): Array<String>

    protected open fun parseKind(kindString: String): Any = kindString
}
